pkg/engine/k8smanifest.go

```diff
@@ -0,0 +1,355 @@
+	vo = addDefaultConfig(vo)
+	// adding ignoreFields from Policy
+	for _, i := range verifyRule.IgnoreFields {
+		converted := k8smanifest.ObjectFieldBinding(i)
+		vo.IgnoreFields = append(vo.IgnoreFields, converted)
+	}
+
+	// dryrun setting
+	vo.DisableDryRun = !verifyRule.DryRunOption.Enable
+	if verifyRule.DryRunOption.Namespace != "" {
+		vo.DryRunNamespace = verifyRule.DryRunOption.Namespace
+	} else {
+		vo.DryRunNamespace = config.KyvernoNamespace()
+	}
+
+	// can be overridden per Attestor
+	if verifyRule.Repository != "" {
+		vo.ResourceBundleRef = verifyRule.Repository
+	}
+
+	// signature annotation
+	// set default annotation domain
+	if verifyRule.AnnotationDomain != "" && verifyRule.AnnotationDomain != DefaultAnnotationKeyDomain {
+		vo.AnnotationConfig.AnnotationKeyDomain = verifyRule.AnnotationDomain
+	}
+
+	// signature verification by each attestor
+	verifiedMsgs := []string{}
+	for i, attestorSet := range verifyRule.Attestors {
+		path := fmt.Sprintf(".attestors[%d]", i)
+		verified, reason, err := verify(resource, attestorSet, vo, path, string(adreq.UID), logger)
+		if err != nil {
+			return verified, reason, err
+		}
+		if !verified {
+			return verified, reason, err
+		} else {
+			verifiedMsgs = append(verifiedMsgs, reason)
+		}
+	}
+	msg := fmt.Sprintf("verified manifest signatures; %s", strings.Join(verifiedMsgs, ","))
+	return true, msg, nil
+}
+
+func verify(resource unstructured.Unstructured, attestorSet kyvernov1.AttestorSet, vo *k8smanifest.VerifyResourceOption, path string, uid string, logger logr.Logger) (bool, string, error) {
+	verifiedCount := 0
+	attestorSet = expandStaticKeys(attestorSet)
+	requiredCount := getRequiredCount(attestorSet)
+	errorList := []error{}
+	verifiedMessageList := []string{}
+	failedMessageList := []string{}
+
+	for i, a := range attestorSet.Entries {
+		var entryError error
+		attestorPath := fmt.Sprintf("%s.entries[%d]", path, i)
+		if a.Attestor != nil {
+			nestedAttestorSet, err := kyvernov1.AttestorSetUnmarshal(a.Attestor)
+			if err != nil {
+				entryError = errors.Wrapf(err, "failed to unmarshal nested attestor %s", attestorPath)
+			} else {
+				attestorPath += ".attestor"
+				verified, reason, err := verify(resource, *nestedAttestorSet, vo, attestorPath, uid, logger)
+				if err != nil {
+					entryError = errors.Wrapf(err, "failed to verify signature; %s", attestorPath)
+				}
+				if verified {
+					// verification success.
+					verifiedCount++
+					verifiedMessageList = append(verifiedMessageList, reason)
+				} else {
+					failedMessageList = append(failedMessageList, reason)
+				}
+			}
+		} else {
+			subPath := ""
+			if a.Keys != nil {
+				subPath = subPath + ".keys"
+				Key := a.Keys.PublicKeys
+				if strings.HasPrefix(Key, "-----BEGIN PUBLIC KEY-----") || strings.HasPrefix(Key, "-----BEGIN PGP PUBLIC KEY BLOCK-----") {
+					// prepare env variable for pubkey
+					pubkeyEnv := fmt.Sprintf("_PK_%s_%d", uid, i)
+					err := os.Setenv(pubkeyEnv, Key)
```

Comment:
Since `os.Setenv` impacts the shared environment state, is this safe for concurrency? 

Is this environment variable used anywhere other than below to create the keypath?

Reply:
I improved this part to use a random number. I think the environment variable is safe because  it set using the admission request ID, key index, and a random number.
Also, I confirmed that this env variable is only used here.